### PR TITLE
feat: config: assert evaluation

### PR DESCRIPTION
**Pull request by katcipis**

### CODE REVIEW

File: docs/codegen/overview.md [hunk 1]
```diff
@@ -8,3 +8,36 @@ each integration scenario in the best way possible. Currently, we support:
 
 * [HCL generation](./generate-hcl.md)
 * [File generation](./generate-file.md)
+
+# Assertions
+
+Assertions can be used in order to fail code generation for one or more stacks
```
Comment (i4ki) [inline docs/codegen/overview.md:14]: it only fails code generation? what about orchestration?

Comment (katcipis) [reply docs/codegen/overview.md:14]: AFAIK it is only code generation (per spec).

File: errors/list_test.go [hunk 2]
```diff
@@ -145,6 +145,34 @@ func TestErrorListFlattensOtherErrorList(t *testing.T) {
...
+	)
+
+	t.Run("empty list match", func(t *testing.T) {
+		assert.IsTrue(t, errors.Is(errors.L(), errors.L()))
+	})
+
+	errs1 := errors.L(errors.E(kind1), errors.E(kind2), errors.E(kind3))
+
+	t.Run("exact errors match", func(t *testing.T) {
+		errs2 := errors.L(errors.E(kind1), errors.E(kind2), errors.E(kind3))
```
Comment (katcipis) [inline errors/list_test.go:162]: Now when we want to check for error lists it should be easier

File: config/assert_test.go [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+				Message:   expr(`"something"`),
+			},
+			want: config.Assert{
+				Assertion: false,
+				Message:   "something",
+			},
+		},
+		{
+			name: "accessing namespace values",
+			namespaces: namespaces{
```
Comment (i4ki) [inline config/assert_test.go:58]: this is cool :heart:

Comment (i4ki) [reply config/assert_test.go:58]: a lot of places could be tested independently of the stack metadata or globals.. what matters is just supports whatever is in the context.

File: config/assert_test.go [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+				Assertion: expr(`unknown.val`),
+				Message:   expr(`false`),
+				Warning:   expr("access.warning"),
+			},
+			wantErr: errors.L(
+				errors.E(eval.ErrEval),
+				errors.E(config.ErrSchema),
+				errors.E(eval.ErrEval),
+			),
+		},
```
Comment (i4ki) [inline config/assert_test.go:169]:
```suggestion
		},
		{
			name: "using funcalls",
			assert: hcl.AssertConfig{
				Assertion: expr(`"A" == tm_upper("a")`),
				Message:   expr(`tm_upper("func")`),
			},
			want: config.Assert{
				Assertion: true,
				Message:   "FUNC",
			},
		},
```
I think a funcall test is missing.

File: config/config.go [hunk 5]
```diff
@@ -27,6 +27,11 @@ const (
 	DefaultFilename = "terramate.tm.hcl"
 )
 
+const (
+	// ErrSchema indicates that the configuration has an invalid schema.
+	ErrSchema errors.Kind = "config has invalid schema"
```
Comment (i4ki) [inline config/config.go:32]:
```suggestion
	ErrSchema errors.Kind = "config has an invalid schema"
```

File: errors/list_test.go (hunk 2 above)
Comment (i4ki) [reply errors/list_test.go:162]: yeah much better, I know the pain, I should have done that in my previous PRs...

**Review by i4ki — APPROVED** [review]
LGTM but I left a test case improvement. Feel free to ignore or address in the following PR if easier.

File: config/assert_test.go (hunk 4 above)
Comment (katcipis) [reply config/assert_test.go:169]: Noice, thanks :100:

**Pull request merged**